ldp/nn/handlers/transformer_handler.py

```diff
@@ -0,0 +1,1007 @@
+    module_call_fn: Callable
+    collate_fn: Callable
+    decollate_fn: Callable
+
+    parallel_mode_config: ParallelModeConfig | None = Field(
+        default=None,
+        description=(
+            "Opt-in config for parallel execution, default of None leads to a"
+            " transformer without parallelism"
+        ),
+    )
+
+    def make_async_module(self, **kwargs) -> AsyncTransformerInterface:
+        if self.parallel_mode_config:
+            return ParallelAsyncTransformer(config=self, **kwargs)
+        return AsyncTransformer(config=self, **kwargs)
+
+
+class AsyncTransformerInterface(ModuleExecutionInterface, AsyncTorchModule, ABC):
+    """Base class for async interactions with a transformer model."""
+
+    @abstractmethod
+    async def __call__(  # type: ignore[override]
+        self,
+        inputs: str | BatchEncoding | list[dict],
+        tools_json: list[dict] | None = None,
+        **kwargs,
+    ) -> tuple[str, torch.Tensor]:
+        """Call the transformer on a single input, which may be encoded."""
+
+    @staticmethod
+    def model_generate(model: PreTrainedModel, *args, **kwargs):
+        """A method that can be used as module_call_fn to sample from an LLM."""
+        # Summoning params per https://github.com/pytorch/pytorch/issues/100069
+        # If model is not FSDP, this context manager is a no-op.
+        with FullyShardedDataParallel.summon_full_params(model, recurse=False):
+            logger.debug(
+                f"model.generate() input_ids shape: {kwargs['input_ids'].shape}, rank {os.environ.get('RANK')}"
+            )
+            return model.generate(
+                *args,
+                **kwargs,
+                pad_token_id=model.config.pad_token_id,  # not always set properly by .generate()
+                # eos_token_id=model.config.eos_token_id,  TODO remove
+            )
+
+
+class TransformerHandler(ModuleHandler):
+    def __init__(self, config: TransformerHandlerConfig):
+        # Maybe this should be configurable? Hard to isolate the effect though
+        torch.set_float32_matmul_precision("high")
+
+        self.config = config
+        # Use local_rank to resolve model location only in the main process *for each node*
+        config.lm_config.resolve_model_location(is_main_process=self.local_rank == 0)
+
+        match config.lm_type:
+            case LMType.GENERATION:
+                tokenizer, model = config.lm_config.get_causal_lm()
+                # On left for https://github.com/huggingface/transformers/pull/7552
+                # ^ that seems to work for most HF models w/ absolute position embeddings
+                # Left padding always works for relative position embeddings
+                tokenizer.padding_side = "left"
+            case LMType.REGRESSION:
+                tokenizer, model = config.lm_config.get_regression_lm()
+            case _:
+                raise NotImplementedError(f"Unsupported LM type: {config.lm_type}")
```

Comment:
```suggestion
                assert_never(config.lm_type)
```

Imo this `NotImplementedError` is a bit misleading because it's not that we don't support it, but it's because the `LMType` enum has no more values. So we can rely on type checker to eliminate this branch for us